lib/rack/request.rb

```diff
@@ -480,25 +480,80 @@ def parseable_data?
         PARSEABLE_DATA_MEDIA_TYPES.include?(media_type)
       end
 
-      # Returns the data received in the query string.
-      def GET
-        if get_header(RACK_REQUEST_QUERY_STRING) == query_string
-          if query_hash = get_header(RACK_REQUEST_QUERY_HASH)
-            return query_hash
+      private def cache_for(key, validity_key, current_value)
+        if get_header(validity_key).equal?(current_value)
+          if has_header?(key)
+            value = get_header(key)
+            if value.is_a?(Exception)
+              raise value.class, value.message, cause: value.cause
+            else
+              return value
+            end
+          end
+        end
+
+        value = yield(current_value)
+
+        set_header(validity_key, current_value)
+        set_header(key, value)
+      rescue => error
+        set_header(key, error)
+        raise
+      end
+
+      # This cache key is used by cached values generated by class_cache_for,
+      # specfically GET and POST. This is to ensure that the cache is not
+      # shared between instances of different classes which have different
+      # behaviour. This includes sub-classes that override query_parser or
+      # expand_params.
+      def cache_key
+        :rack
+      end
+
+      private def class_cache_for(key, validity_key, current_value)
+        cache_key = self.cache_key
+
+        if get_header(validity_key).equal?(current_value)
+          if cache = get_header(key)
+            if cache.key?(cache_key)
+              value = cache[cache_key]
+              if value.is_a?(Exception)
+                raise value.class, value.message, cause: value.cause
+              else
+                return value
+              end
+            end
           end
         end
 
-        set_header(RACK_REQUEST_QUERY_HASH, expand_params(query_param_list))
+        unless cache
+          set_header(key, cache = {})
+        end
+
+        cache.fetch(cache_key) do
+          value = yield(current_value)
+
+          # Only set this after generating the value, so that if an error or other cache depending on the same key, it will be invalidated correctly:
+          set_header(validity_key, current_value)
+
+          return cache[cache_key] = value
+        rescue => error
+          cache[cache_key] = error
```

Comment:
Can we get some comment around here.  I see it's just moved from `rack request form error`, but caching an exception makes me raise my eyebrows.  It would be helpful to have context around this code.

Reply:
The cached operation can fail, and if it fails, subsequent calls to the cache should also fail with the same exception. An example of this is missing input, failed multipart parse, etc.

This was discussed a bit in the related PR: https://github.com/rack/rack/pull/2010

This is covered by tests already. Those tests are still passing in this PR.